src/transaction/input/mod.rs

```diff
@@ -0,0 +1,107 @@
+pub mod transaction_input;
+use std::str::FromStr;
+
+use ckb_types::packed;
+pub use transaction_input::TransactionInput;
+
+use crate::{
+    traits::{
+        CellCollector, CellCollectorError, CellQueryOptions, DefaultCellCollector, ValueRangeOption,
+    },
+    types::NetworkInfo,
+    Address,
+};
+
+pub struct InputIterator {
+    pub(crate) buffer_inputs: Vec<TransactionInput>,
+    pub(crate) buffer_index: usize,
+    pub(crate) script_index: usize,
+    pub(crate) lock_scripts: Vec<packed::Script>,
+    pub(crate) cell_collector: DefaultCellCollector,
+}
+
+impl InputIterator {
+    pub fn new(lock_scripts: Vec<packed::Script>, network_info: &NetworkInfo) -> Self {
+        Self {
+            buffer_inputs: vec![],
+            buffer_index: 0,
+            script_index: 0,
+            lock_scripts,
+            cell_collector: DefaultCellCollector::new(&network_info.url),
+        }
+    }
+    pub fn new_with_address<T: AsRef<str>>(
+        address: &[T],
+        network_info: &NetworkInfo,
+    ) -> Result<Self, String> {
+        let lock_sripts: Result<Vec<packed::Script>, String> = address
+            .iter()
+            .map(|adr_str| Address::from_str(adr_str.as_ref()))
+            .map(|ad_r| ad_r.map(|t| Into::<packed::Script>::into(&t)))
+            .collect();
+        let lock_scripts = lock_sripts?;
+        Ok(Self::new(lock_scripts, network_info))
+    }
+
+    fn collect_live_cells_by_lock(
+        cell_collector: &mut DefaultCellCollector,
+        buffer_inputs: &mut Vec<TransactionInput>,
+        buffer_index: &mut usize,
```

Comment:
While use script_index, the script is a shared reference, then can not use a "mut self" while hold part of the object;

After use pop, the script is not a share reference anymore, then it can with mut self.